azure/services/aso/aso.go

```diff
@@ -0,0 +1,165 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aso
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/conditions"
+	"github.com/google/go-cmp/cmp"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Service is an implementation of the Reconciler interface. It handles creation
+// and deletion of resources using ASO.
+type Service struct {
+	client.Client
+}
+
+// New creates a new ASO service.
+func New(ctrlClient client.Client) *Service {
+	return &Service{
+		Client: ctrlClient,
+	}
+}
+
+// CreateOrUpdateResource implements the logic for creating a new or updating an
+// existing resource with ASO.
+func (s *Service) CreateOrUpdateResource(ctx context.Context, spec azure.ASOResourceSpecGetter, serviceName string) (result genruntime.MetaObject, err error) {
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "aso.Service.CreateOrUpdateResource")
+	defer done()
+
+	resource := spec.ResourceRef()
+	resourceName := resource.GetName()
+	resourceNamespace := resource.GetNamespace()
+
+	log = log.WithValues("service", serviceName, "resource", resourceName, "namespace", resourceNamespace)
+
+	var existing genruntime.MetaObject
+	if err := s.Client.Get(ctx, client.ObjectKeyFromObject(resource), resource); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, errors.Wrapf(err, "failed to get existing resource %s/%s (service: %s)", resourceNamespace, resourceName, serviceName)
+		}
+	} else {
+		existing = resource
+		log.V(2).Info("successfully got existing resource")
+
+		// Check if there is an ongoing long running operation.
+		conds := existing.GetConditions()
+		i, readyExists := conds.FindIndexByType(conditions.ConditionTypeReady)
+		if !readyExists {
+			return nil, azure.WithTransientError(errors.New("ready status unknown"), 20*time.Second)
+		}
+		var readyErr error
+		if cond := conds[i]; cond.Status != metav1.ConditionTrue {
+			switch {
+			case cond.Reason == conditions.ReasonReconciling.Name:
+				readyErr = newOperationNotDoneError(existing)
+			default:
+				readyErr = fmt.Errorf("resource is not Ready: %s", conds[i].Message)
+			}
+
+			if readyErr != nil {
+				if conds[i].Severity == conditions.ConditionSeverityError {
```

Comment:
For my own understanding: if the condition has the severity Error, that means it's always a terminal error? If so, how are transient errors distinguished if there is not condition for it?

Reply:
My read of the [ASO docs](https://azure.github.io/azure-service-operator/guide/conditions/#severity) says `Error` should be terminal and any other severity is treated as transient. In either case, we will bubble up the ASO condition's `message` which might provide more context.